### PR TITLE
Updated Qt-OpenGL Camera

### PR DESCRIPTION
**Pull request by allsey87**
This pull request needs to be reviewed and further tested before merge. It is a significant overhaul of the Qt-OpenGL camera. Among many clean ups and refactoring the new camera now supports:
1. Specifying a timeline with optional interpolation and looping
2. Setting camera placements to sensible defaults with respect to the arena size and center

### CODE REVIEW

**Comment by allsey87** [comment]
For an example of using the new camera, please see [dynamics3d_single_body_test.argos](https://github.com/ilpincy/argos3/blob/qtopengl-new-camera/src/testing/examples/dynamics3d_single_body_test.argos)

@jharwell could you also give me your feedback on these changes?

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 1]
```diff
@@ -67,308 +64,207 @@ namespace argos {
...
    /****************************************/
 
    void CQTOpenGLCamera::Init(TConfigurationNode& t_tree) {
-      if(NodeExists(t_tree, "camera")) {
-         try {
-            TConfigurationNode tCameraNode;
-            tCameraNode = GetNode(t_tree, "camera");
-            /* Parse "placement" nodes */
+      try {
+         if(NodeExists(t_tree, "placements")) {
```
Comment (jharwell) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:113]: This function seems too long. Consider breaking into smaller functions for improved readability:

- Parsing the placements
- Parsing the timeline nodes

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 2]
```diff
@@ -377,33 +273,161 @@ namespace argos {
...
-         /* Yes, save it and search for subsequent items */
-         STimelineItem* psTimelineItem = *it;
-         ++it;
-         while(it != m_listTimeline.end() && (*it)->Step <= unStep) {
-            psTimelineItem = *it;
-            ++it;
+      if(!m_bEnableTimeline)
+         return;
+      /* initialize two iterators to the first keyframe */
+      std::vector<SKeyframe>::const_iterator itCurrent =
```
Comment (jharwell) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:335]: This function seems too long. Consider breaking into smaller functions to improve readability:

- Handle looping case
- Handle non-looping case

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h [hunk 3]
```diff
@@ -37,85 +37,54 @@ namespace argos {
...
-         /** Calculate the sensitivity of the camera */
-         void CalculateSensitivity();
       };
-
-      struct STimelineItem {
-         /** Index of the placement */
-         UInt32 Idx;
-         /** Time step to switch to the indicated placement */
+ 
+      struct SKeyframe {
```
Comment (jharwell) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h:54]: Missing documentation. This is an important data structure and it should be documented to help demystify usage

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 4]
```diff
@@ -73,7 +73,31 @@ namespace argos {
...
+         cAngle *= static_cast<Real>(un_index);
+         m_arrPlacements[un_index].Position = 
+            CVector3::X * cArenaSize.Length();
+         m_arrPlacements[un_index].Position.RotateZ(cAngle);
+         m_arrPlacements[un_index].Position += cArenaCenter;
+         m_arrPlacements[un_index].Target = cArenaCenter;
+         m_arrPlacements[un_index].Target.SetZ(0);
+         m_arrPlacements[un_index].Up = 
+            /* position - target? */
+            CVector3(0.0, 0.0, 1.0).Normalize();
```
Comment (jharwell) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:97]: Use CVector3::Z for readability (I think that exists?)

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 5]
```diff
@@ -73,7 +73,31 @@ namespace argos {
...
+         m_arrPlacements[un_index].LensFocalLength = 0.08;
```
Comment (jharwell) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:98]: Magic number

File: src/testing/examples/dynamics3d_single_body_test.argos [hunk 6]
```diff
@@ -64,17 +64,27 @@
...
-        <placement idx="1" position=" 1.25,-1.25,0.6" look_at="0,0,0.05" lens_focal_length="92" />
-        <placement idx="2" position=" 1.25, 0,   0.5" look_at="0,0,0.05" lens_focal_length="78" />
-        <placement idx="3" position=" 1.25, 1.25,0.6" look_at="0,0,0.05" lens_focal_length="92" />
-        <placement idx="4" position=" 0,    1.25,0.5" look_at="0,0,0.05" lens_focal_length="78" />
-        <placement idx="5" position="-1.25, 1.25,0.6" look_at="0,0,0.05" lens_focal_length="92" />
-        <placement idx="6" position="-1.25, 0,   0.5" look_at="0,0,0.05" lens_focal_length="78" />
-        <placement idx="7" position="-1.25,-1.25,0.6" look_at="0,0,0.05" lens_focal_length="92" />
-        <placement idx="8" position="0,0,0.5" look_at="0.0,0.0,0.125" up="1,0,0" lens_focal_length="40"/>
+        <placements>
+          <placement index="0" position=" 0,   -1.25,0.5" look_at="0,0,0.05" lens_focal_length="78" />
```
Comment (jharwell) [inline src/testing/examples/dynamics3d_single_body_test.argos:70]: Example looks good--this should be used to update the results of "argos3 -q opengl" with the new functionality

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 7]
```diff
@@ -13,8 +13,8 @@
 
 namespace argos {
 
-   static const Real MOVE_GAIN = 0.005f / ::exp(0.02f);
-   static const Real ROTATE_GAIN = 0.01f / ::exp(-0.02f);
+   static const Real MOVE_GAIN = 0.005 / Exp(0.02);
```
Comment (jharwell) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:16]: Magic numbers need documentation

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 8]
```diff
@@ -23,41 +23,38 @@ namespace argos {
...
          GetNodeAttribute(t_tree, "up", Up);
          Up.Normalize();
-         /* Calculate the Left vector with a cross product */
-         Left = Up;
-         Left.CrossProduct(Forward).Normalize();
       }
-      /* Get optional optics parameters */
+      /* Get focal length */
       Real fValue;
       GetNodeAttributeOrDefault<Real>(t_tree, "lens_focal_length", fValue, 20.0f);
```
Comment (jharwell) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:59]: Magic number

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 9]
```diff
@@ -67,308 +64,207 @@ namespace argos {
...
+               cAngle *= static_cast<Real>(un_index);
+               m_arrPlacements[un_index].Position = 
+                  CVector3::X * cArenaSize.Length();
+               m_arrPlacements[un_index].Position.RotateZ(cAngle);
+               m_arrPlacements[un_index].Position += cArenaCenter;
+               m_arrPlacements[un_index].Target = cArenaCenter;
+               m_arrPlacements[un_index].Target.SetZ(0);
+               m_arrPlacements[un_index].Up = 
+                  /* position - target? */
+                  CVector3(0.0, 0.0, 1.0).Normalize();
```
Comment (jharwell) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp]: Use CVector3::Z

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 1 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:113]: I wouldn't break this into separate functions. I think it is a bit of an anti-pattern to break a function down into smaller functions when they are only going to be called from a exactly one call site.

Comment (ilpincy) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:113]: I agree, function length alone is not a good metric to make the structure more complex.

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 4 above)
Comment (ilpincy) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:97]: Also, why would you normalize a vector that is created already normalized?

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 2 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:335]: This function is too long and I think there are better ways of handling this logic. Although for this PR, I am focusing mostly on correctness and functionality. I will leave further refactoring and improving this logic to a future PR...

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 10]
```diff
@@ -23,41 +23,38 @@ namespace argos {
...
-      Forward = (Target - Position).Normalize();
       /* Was the up vector specified? */
       if(!NodeAttributeExists(t_tree, "up")) {
          /* Calculate the Left vector
             It is located on a plane parallel to XY
             It is perpendicular to the projection of Forward on that plane
          */
-         if(Forward.GetX() != 0 || Forward.GetY() != 0) {
-            CVector2 cLeftXY(Forward.GetX(), Forward.GetY());
+         CVector3 cForward((Target - Position).Normalize());
```
Comment (ilpincy) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:32]: This code is wrong. If the difference between `Target` and `Position` is zero, you'll get a `NaN`.

Comment (ilpincy) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:32]: I notice now that also the original code had the same bug... Oh well :)

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 11]
```diff
@@ -23,41 +23,38 @@ namespace argos {
...
+         CVector3 cLeft;
```
Comment (ilpincy) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:33]: Why did you remove the `Left` vector from the struct and recalculated it here privately? Having the `Left` vector available is a good idea for programmatically changing the camera code in the `QtUserFunctions`, in case people would like to have that possibility.

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 12]
```diff
@@ -67,308 +64,213 @@ namespace argos {
...
-         unchanged!
-      */
-      Left = Up;
-      Left.CrossProduct(Forward).Normalize();
-   }
-
-   /****************************************/
-   /****************************************/
-
-   void CQTOpenGLCamera::SPlacement::RotateLeftRight2(const CRadians& c_angle) {
```
Comment (ilpincy) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:157]: Why did you remove this code?

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h [hunk 13]
```diff
@@ -37,85 +37,54 @@ namespace argos {
...
          CVector3 Up;
-         /** The local X axis of the camera in the global reference frame */
-         CVector3 Forward;
-         /** The direction of sight of the camera in the global reference frame */
-         CVector3 Target;
          /** The focal length of the lens (if this was a real camera) */
          Real LensFocalLength;
          /** The focal length of the camera */
          CDegrees YFieldOfView;
-         /** Motion sensitivity */
```
Comment (ilpincy) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h:52]: Why did you remove this?

File: src/testing/examples/dynamics3d_single_body_test.argos (hunk 6 above)
Comment (allsey87) [reply src/testing/examples/dynamics3d_single_body_test.argos:70]: I will add this to the to-do list

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h [hunk 14]
```diff
@@ -37,85 +37,54 @@ namespace argos {
       struct SPlacement {
          /** The position of the camera in the global reference frame */
          CVector3 Position;
-         /** The local Y axis of the camera in the global reference frame */
```
Comment (ilpincy) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h:40]: Why did you remove this?

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 7 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:16]: I leave this to a future clean up

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h [hunk 15]
```diff
@@ -37,85 +37,54 @@ namespace argos {
...
-            Target(),
-            LensFocalLength(0.02f),
-            MotionSensitivity(0.005),
-            RotationSensitivity(0.01) {
-            CalculateYFieldOfView();
-         }
-
          /** Initialize from XML */
          void Init(TConfigurationNode& t_tree);
-         /** Rotation around the local Y axis */
```
Comment (ilpincy) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h:71]: Why did you remove these?

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 8 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:59]: I leave this to a future clean up

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 11 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:33]: The problem is related to the interpolation. It doesn't make sense to constantly recalculate these vectors unless the user want to move the camera via user input

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 16]
```diff
@@ -67,308 +64,213 @@ namespace argos {
...
-         .Target += m_sPlacement[m_unActivePlacement].Forward;
+      /* disable timeline on rotate */
+      m_bEnableTimeline = false;
+      /* Calculate cForward and cLeft */
+      CVector3 cForward((m_sActivePlacement.Target - m_sActivePlacement.Position).Normalize());
+      CVector3 cLeft = m_sActivePlacement.Up;
+      cLeft.CrossProduct(cForward).Normalize();
+      /* Calculate fRotationSensitivity */
+      Real fRotationSensitivity = 
+         ROTATE_GAIN * ::exp(-m_sActivePlacement.LensFocalLength);
```
[the remark this answers is not in the document]
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:219]: I was removing them, I missed this one

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 12 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:157]: It has been refactored into `CQTOpenGLCamera::Rotate`

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h (hunk 13 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h:52]: It is only needed during `move`/`rotate` so it is now calculated on demand. It doesn't make sense to recalculate it every time active camera placement is updated during interpolate.

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 11 above)
Comment (ilpincy) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:33]: I don't think I understand :( Having a `Left` vector in the struct is the same in terms of when to compute it, but it makes it much simpler and faster to rotate around that vector instead of having to recalculate it.

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h (hunk 14 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h:40]: Same reason as motion sensitivity

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h (hunk 15 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h:71]: Same reason as motion sensitivity

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 10 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:32]: I will check this, but I do not think I change the logic of the code only refactored it a bit...

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 5 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:98]: I agree that magic numbers are no good, but these already existed in the old version and I don't have time right now to fix them.

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp [hunk 17]
```diff
@@ -67,308 +64,213 @@ namespace argos {
...
+                cTarget.GetZ(),
+                cUp.GetX(),
+                cUp.GetY(),
+                cUp.GetZ());
    }
 
    /****************************************/
    /****************************************/
 
    void CQTOpenGLCamera::Rotate(const QPoint& c_delta) {
```
Comment (ilpincy) [inline src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:210]: Why is rotate accepting a `QPoint`? A rotation should expect an angle...

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 4 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp]: Fixed in 1588d020fc01cde3cc3ef7750bbbb8b200b85b44

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h (hunk 3 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.h:54]: Fixed in 1588d020fc01cde3cc3ef7750bbbb8b200b85b44

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 9 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp]: Fixed in 1588d020fc01cde3cc3ef7750bbbb8b200b85b44

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp]: Fixed in 1588d020fc01cde3cc3ef7750bbbb8b200b85b44

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 12 above)
Comment (ilpincy) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:157]: Why? At least, please keep the comments I wrote. It's not clear what the new `Rotate()` function does, partly because of the missing comments, partly because instead of accepting an angle as input, it accepts a `QPoint` (and then it immediately calculates an angle?).

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 17 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:209]: @ilpincy that was how it was when I started. The code has just just refactored, the logic should not have changed...

**Comment by allsey87** [comment]
@ilpincy @jharwell This class was and still is a bit of a mess in my opinion. However, I believe it is already much better than it was before. Personally, I don't think it worth investing more time into this since (i) it seems at least as correct as it was before and (ii) @ilpincy has mentioned numerous times in the past the thought of replacing this entire plugin anyway....

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 17 above)
Comment (ilpincy) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:209]: The original function was meant to be a convenience function that used simpler functions to do its job. The real work is done in the simpler functions. I prefer the original structure because it's more readable.

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 12 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:157]: I have a note and a couple opinions here:
1. `CQTOpenGLCamera::Rotate` always took a `QPoint` as a single argument
2. All of the rotate function functions were only ever called from `CQTOpenGLCamera::Rotate` and only served the purpose of rotating the camera with respect to the user's mouse
3.  The comments in these functions were far too verbose and made the code hard to follow, perhaps such documentation should be on the website or a wiki?

Anyway, I already spent too much time working on this so if adding the comments back in and spiting the functions up is what it takes to get this merged, I will do it

Comment (ilpincy) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:157]: 1. This is a convenience function for the Qt visualization.
2. Those functions make it possible to programmatically rotate the camera without reference to the mouse. I used them in the past to program panning in my videos. Plus, they're simple to read.
3. When I made that code, it took me 10 hours to get it right - it was quite a process for me. I don't think it's bad if the comments are verbose. I prefer to use a little bit of literate programming and not have to remember the equations, rather than having to guess them from the code.

I think it's not a problem if you stop working on this now. I'll have a thorough look at this code after August 11, and I'll probably changes things around anyway.

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 11 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:33]: So we now have a `m_sActivePlacement` that represents the current placement of the camera (including in between two placements from the array `m_arrPlacements` during interpolation). It is true, that we could precalculate several things inside `SPlacement` which would give a minor performance boost while rotating or moving the camera with the mouse. On the other hand, we now need to recalculate these vectors on each time step for `m_sActivePlacement` during interpolation to keep it in a correct state even though we are probably not going to use them. Removing this additional data from `SPlacement` may also bring a minor performance increase since the overall size of `m_arrPlacements` is smaller.

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 17 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:209]: As I mentioned in another thread on this PR, I can revert this back to be more similar to how it was before.

Comment (ilpincy) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:209]: Feel free to go ahead! I'm just saying that I'll give this PR a thorough look on August 11 (working on a proposal) and I might change a couple of things, so the work you did is already a lot. I don't want to make you waste more time for aspects I could fix myself with a bit of copy-paste.

**Comment by allsey87** [comment]
So I will implement the following changes before pulling this into master
1. Update the documentation
2. Restore comments and functions related to rotation
3. Update the remaining .argos files in the testing directory

I will then leave the final tweaks, clean ups, optimizations to @ilpincy to do on the 11th of August

**Comment by ilpincy** [comment]
OK for the changes, but I would prefer if you waited before pulling into master. I didn't have time to really check the code yet, I just looked at the most evident changes. Is this urgent?

**Comment by allsey87** [comment]
I can wait a bit, however, I would like this work to be pulled into master by 15 August at the absolute latest...

**Comment by ilpincy** [comment]
If it's urgent, go ahead. The thing is that, in case of bugs, the requests always end up on my lap, especially on the forum. I'd prefer to wait and be sure the code works in a way I understand before merging. If you're confident the code works and can commit to checking the forum regularly, go ahead. I'll change things at a later time.

**Comment by allsey87** [comment]
I will implement the aforementioned features and continue testing before I pull it into master. It will be hard to say I got everything without some form of fuzzing but I believe I have covered all of the corner cases. I am committed to answering questions on the forum as soon as you fix it to send me notifications about new questions etc.

**Comment by ilpincy** [comment]
The forum needs fixing? I receive emails and questions routinely, what is the issue?

**Comment by allsey87** [comment]
I think I have mentioned this a couple times over the past two years... For my account on forum, I do not get notifications and do not have an option to turn them on.

**Comment by ilpincy** [comment]
I checked the account. The associated e-mail was from Paderborn, I changed to ULB and made you an admin.

**Comment by allsey87** [comment]
Thanks, I was able to login. Let see if I now receive notifications about new posts...

File: src/testing/examples/dynamics3d_single_body_test.argos (hunk 6 above)
Comment (allsey87) [reply src/testing/examples/dynamics3d_single_body_test.argos:70]: Fixed in e3b31e21f310eac24ca0210a066f1fd9e73982c6

File: src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp (hunk 12 above)
Comment (allsey87) [reply src/plugins/simulator/visualizations/qt-opengl/qtopengl_camera.cpp:157]: Fixed in 9e922e763ec1d0a0ed639f0f5880a2a37d575623

**Comment by allsey87** [comment]
These commits over the past 10 days address the issues previously raised in this PR. @ilpincy since it is the 11th of August (when you mentioned you are free to look over this), I will let you do that before I merge the code tomorrow. Feel free to change/improve anything you want, I will try to reply here promptly should you have any questions :)

**Comment by ilpincy** [comment]
I said after August 11th :-) Today is the deadline of my proposal. I'll have a look at the PR this week.

**Comment by allsey87** [comment]
No worries @ilpincy, best of luck with getting that proposal submitted :)

**Comment by allsey87** [comment]
So, can I get the go ahead to merge this into master? :)

**Comment by allsey87** [comment]
@ilpincy if you can't have something done by a particular date, it would be appreciated if you could reply here and give another estimate for when it can be done. Just ignoring these messages is frustrating to your collaborators at the least.

**Comment by ilpincy** [comment]
I am sorry. I am doing the best I can juggling many tasks, some of which were urgent, stressful and unplanned. I'll just merge this without checking it and if there's bugs I'll send people your way.
